File: .github/scripts/check_pr.js
import fs from 'fs';
import path from 'path';
import { spawnSync } from 'child_process';
import { fileURLToPath } from 'url';
import yaml from 'js-yaml';
import Ajv from 'ajv';
import addFormats from 'ajv-formats';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const ajv = new Ajv();
addFormats(ajv);

function checkNoExternalLinks(htmlFile) {
  const content = fs.readFileSync(htmlFile, 'utf8');
  // allowed external domains
  const sanitized = content
    //  SVG/XML namespaces
    .replace(/xmlns=['"]http:\/\/www\.w3\.org\/2000\/svg['"]/g, '')
    .replace(/xmlns=['"]http:\/\/www\.w3\.org\/1999\/xhtml['"]/g, '')
    // Unsplash images
    .replace(/https:\/\/images\.unsplash\.com\/?/g, '')
    // Google Fonts
    .replace(/https:\/\/fonts\.googleapis\.com\/?/g, '')
    .replace(/https:\/\/fonts\.gstatic\.com\/?/g, '');

  return !sanitized.includes('http://') && !sanitized.includes('https://');
}

function main() {
  const repoRoot = path.resolve(__dirname, '..', '..');
  const designMdDir = path.join(repoRoot, 'design-md');
  const schemaPath = path.join(repoRoot, 'metadata.schema.yaml');

  if (!fs.existsSync(schemaPath)) {
    console.error(`Schema file not found at ${schemaPath}`);
    process.exit(1);
  }

  const schema = yaml.load(fs.readFileSync(schemaPath, 'utf8'));
  const validate = ajv.compile(schema);

  const errors = [];

  const entries = fs.readdirSync(designMdDir);

  for (const entry of entries) {
    const entryPath = path.join(designMdDir, entry);
    if (!fs.statSync(entryPath).isDirectory()) continue;
    if (entry.startsWith('.')) continue;

    const designFile = path.join(entryPath, 'DESIGN.md');
    const previewHtml = path.join(entryPath, 'preview.html');
    const previewPng = path.join(entryPath, 'preview.png');
    const metadataYaml = path.join(entryPath, 'metadata.yaml');

    if (!fs.existsSync(designFile)) continue;

    console.log(`Checking ${entry}...`);

    // 1. Linter
    const lintRes = spawnSync('npx', ['@google/design.md', 'lint', 'DESIGN.md'], {
      cwd: entryPath,
      encoding: 'utf8',
      shell: true
    });
    if (lintRes.status !== 0) {
      errors.push(`${entry}: Linter failed:\n${lintRes.stdout}\n${lintRes.stderr}`);
    }

    // 2. preview.html exists
    if (!fs.existsSync(previewHtml)) {
      errors.push(`${entry}: preview.html is missing`);
    } else {
      // 3. No external links
      if (!checkNoExternalLinks(previewHtml)) {
        errors.push(`${entry}: preview.html contains external links (http:// or https://) which are forbidden`);
      }
    }

    // 4. preview.png exists
    if (!fs.existsSync(previewPng)) {
      errors.push(`${entry}: preview.png is missing`);
    }

    // 5. metadata correct
    if (!fs.existsSync(metadataYaml)) {
      errors.push(`${entry}: metadata.yaml is missing`);
    } else {
      try {
        const metadata = yaml.load(fs.readFileSync(metadataYaml, 'utf8'));
        const valid = validate(metadata);
        if (!valid) {
          errors.push(`${entry}: metadata.yaml validation failed: ${ajv.errorsText(validate.errors)}`);
        }
      } catch (e) {
        errors.push(`${entry}: metadata.yaml parse error: ${e.message}`);
      }
    }
  }

  if (errors.length > 0) {
    console.error("\nValidation errors found:");
    errors.forEach(e => console.error(`- ${e}`));
    process.exit(1);
  } else {
    console.log("\nAll checks passed.");
  }
}

main();

File: .github/scripts/check_pr.py
import os
import sys
import subprocess
import yaml
import jsonschema

def check_no_external_links(html_file):
    with open(html_file, 'r', encoding='utf-8') as f:
        content = f.read()
        if 'http://' in content or 'https://' in content:
            return False
    return True

def main():
    repo_root = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
    design_md_dir = os.path.join(repo_root, 'design-md')
    schema_path = os.path.join(repo_root, 'metadata.schema.yaml')
    
    with open(schema_path, 'r', encoding='utf-8') as f:
        schema = yaml.safe_load(f)

    errors = []

    for entry in os.listdir(design_md_dir):
        entry_path = os.path.join(design_md_dir, entry)
        if not os.path.isdir(entry_path):
            continue
            
        design_file = os.path.join(entry_path, 'DESIGN.md')
        preview_html = os.path.join(entry_path, 'preview.html')
        preview_png = os.path.join(entry_path, 'preview.png')
        metadata_yaml = os.path.join(entry_path, 'metadata.yaml')

        if not os.path.exists(design_file):
            continue
            
        # 1. Linter
        res = subprocess.run(['npx', '@google/design.md', 'lint', 'DESIGN.md'], cwd=entry_path, capture_output=True, text=True)
        if res.returncode != 0:
            errors.append(f"{entry}: Linter failed:\n{res.stdout}\n{res.stderr}")

        # 3. preview html is available
        if not os.path.exists(preview_html):
            errors.append(f"{entry}: preview.html is missing")
        else:
            # 2. No links in preview
            if not check_no_external_links(preview_html):
                errors.append(f"{entry}: preview.html contains external links (http:// or https://) which are forbidden")

        # 4. preview image is available
        if not os.path.exists(preview_png):
            errors.append(f"{entry}: preview.png is missing")

        # 5. metadata correct
        if not os.path.exists(metadata_yaml):
            errors.append(f"{entry}: metadata.yaml is missing")
        else:
            try:
                with open(metadata_yaml, 'r', encoding='utf-8') as f:
                    metadata = yaml.safe_load(f)
                jsonschema.validate(instance=metadata, schema=schema)
            except Exception as e:
                errors.append(f"{entry}: metadata.yaml validation failed: {str(e)}")

    if errors:
        print("Validation errors found:")
        for e in errors:
            print(f"- {e}")
        sys.exit(1)
    else:
        print("All checks passed.")

if __name__ == '__main__':
    main()
